### DISCUSSION TITLE
建议拆分倍速扩展功能和倍速记忆功能

### DISCUSSION THREAD

**Discussion by ZiSaAcZ7** [the1812/Bilibili-Evolved#2623] [功能建议 / Ideas]
我个人对倍速扩展功能很喜欢，经常3倍速看视频，但是我并不想要倍速记忆功能。
现在倍速记忆功能似乎是倍速扩展功能的父功能，关闭记忆功能就无法使用倍速扩展功能。

**Comment by LonelySteve** [comment]
你说的很有道理，其实我最初设计的时候就想这么做，但因为某些原因，之前 v1 版的实现是耦合在一起的，没法分开。

本来打算迁到 v2 版的时候把它重构一下分开的，但是 v2 出 TP 那段时间我比较忙，@the1812 就帮我简单粗暴地迁移过去了，考虑到已经有用户在使用 v2 版了，如果此时分离功能会导致现在用户的配置不可用。

因此我大概需要**本体**提供一种标记组件被废弃的功能：

- 在组件详情面板展示被废弃的组件的废弃原因和解决办法
- 在用户尝试安装被废弃的组件时，弹出提示
- 引导用户安装新组件后，自动卸载被废弃的组件，并能够迁移废弃组件的配置

目前 v2 版的倍速记忆功能已经被我重构了一次，从理论上来说，独立出倍速扩展和倍速记忆功能几乎肯定是可以做到的。

待与 @the1812 进一步讨论之后，我再确定下一步该怎么做。

**Reply by the1812** [reply]
为什么会有废弃的组件? 拆分后原组件应该是倍速记忆和倍速扩展其中的一个吧

**Reply by LonelySteve** [reply]
我这么想的原因是，是因为我不想将原有组件拆分成两个组件，而是替换为另外一个（这样改动比较小），只是菜单发生变化，变成：

- 倍速增强 > 倍速记忆
- 倍速增强 > 倍速扩展

好吧，可能还是拆成两个不相关的组件会更好，或许目前来说，不需要做废弃功能，但那些具有插件系统的软件基本都有个废弃功能支持。。。

那么，我主要还有两个问题：

1. 如何迁移现有配置呢？要知道，现在扩展倍速值是放在 rememberVideo 配置下的，新分离的组件最好能通过某种生命周期或者钩子之类的东西（postInstall）告诉新组件自己被安装了，这样新组件就能够迁移旧组件的配置（如果有旧组件的话）
2. 用户主动升级“倍速记忆”组件后，因原有倍速扩展组件功能分离，对于用户来说，感觉就像下线了这个功能一样，肯定有用户会跑到 issues 区提问，即使不采取废弃机制，那也最好有个方式告知用户

**Reply by the1812** [reply]
我问这个问题是顺着你的废弃组件的想法来问的, 其实我自己一开始想的也是一个组件内提供两个分离功能的选项开关 (类似现在的 `签到助手`), 这样只需要新增一个 "是否记忆倍速" 之类的选项, 不过 componentName 就还是 `rememberVideoSpeed` 可能有点怪.

如果是拆组件, 旧组件按名称 `rememberVideoSpeed` 应该是 `倍速记忆`, 拆出去一个 `倍速扩展` 组件.
1. 设置的迁移, 可以在 `entry` 里检测 `rememberVideoSpeed` 的 `options`, 若有就执行迁移, 然后删除 `rememberVideoSpeed` 那边的相应 `options`, 这样之后就不会执行迁移代码, 其实和 `postInstall` 的行为是一样的.
2. 组件的 `description` 支持 Markdown, 直接写上相应公告即可, 如果不放心, 甚至还可以像上面一样, 检测到未迁移的的 `options` 时, 弹一个 Toast 告知用户功能已分离, 用户可以选择:
  - 不使用 `倍速扩展`, 直接删除未迁移的 `options`
  - 使用 `倍速扩展`, 安装后 `倍速扩展` 的 `entry` 如 1 中所述完成迁移后删除 `options`

**Reply by LonelySteve** [reply]
> 我问这个问题是顺着你的废弃组件的想法来问的, 其实我自己一开始想的也是一个组件内提供两个分离功能的选项开关 (类似现在的 签到助手), 这样只需要新增一个 "是否记忆倍速" 之类的选项, 不过 componentName 就还是 `rememberVideoSpeed` 可能有点怪.

是啊，所以我才想要废弃 `rememberVideoSpeed` 😂

> 如果是拆组件, 旧组件按名称 `rememberVideoSpeed` 应该是 `倍速记忆`, 拆出去一个 `倍速扩展` 组件.
> 
> 1. 设置的迁移, 可以在 `entry` 里检测 `rememberVideoSpeed` 的 `options`, 若有就执行迁移, 然后删除 `rememberVideoSpeed` 那边的相应 `options`, 这样之后就不会执行迁移代码, 其实和 `postInstall` 的行为是一样的.
> 2. 组件的 `description` 支持 Markdown, 直接写上相应公告即可, 如果不放心, 甚至还可以像上面一样, 检测到未迁移的的 `options` 时, 弹一个 Toast 告知用户功能已分离, 用户可以选择:
> 
> * 不使用 `倍速扩展`, 直接删除未迁移的 `options`
> * 使用 `倍速扩展`, 安装后 `倍速扩展` 的 `entry` 如 1 中所述完成迁移后删除 `options`

妙啊妙啊，我懂啦！

**Reply by deathwood** [reply]
现在必须开了倍数记忆才能用倍数的快捷键么（shift +上下），想要他们分开

**Reply by LonelySteve** [reply]
> 现在必须开了倍数记忆才能用倍数的快捷键么（shift +上下），想要他们分开

对于快捷键扩展，或许可以考虑做成一个独立于组件的插件，等我研究一下，应该可以做到只需安装`倍速记忆`或者`倍速扩展`其中一个就可以起作用的。

**Comment by ljcbaby** [comment]
倍数扩展后可能缓冲机制也要调整，不然会看一段缓冲一段

**Reply by FantasticAiming** [reply]
我一直以为是我网络的问题..

**Comment by zhj9709** [comment]
我现在只有开了倍速记忆，快捷键调倍速才有用

**Reply by LonelySteve** [reply]
这个问题会在新版解决，新的版本将分为两个组件，一个插件。两个组件分别实现「倍速记忆」和「倍速扩展」，而插件实现快捷键支持。。

这样大家就可以各取所需了，可惜最近太忙了，抽不出时间弄这个，否则上个月就应该可以发布的。😥
